ree/reeCorrectedFlux.cc

```diff
@@ -0,0 +1,299 @@
+
+    for (int i = 0; i < np; i++) {
+      apf::setComponents(cf->correctedFlux, tet, i, zeros);
+    }
+  }
+  cf->mesh->end(it);
+}
+
+typedef std::vector<apf::MeshEntity*> EntityVector;
+struct FaceCavity
+{
+  apf::Mesh* mesh;
+  apf::MeshEntity* entity;
+  CorrectFlux* correctflux;
+  EntityVector tets;
+};
+
+static void setupFaceCavity(FaceCavity* fc, CorrectFlux* cf)
+{
+  fc->mesh = cf->mesh;
+  fc->correctflux = cf;
+  fc->entity = 0;
+}
+
+static void startFaceCavity(FaceCavity* fc, apf::MeshEntity* f)
+{
+  fc->entity = f;
+  fc->tets.clear();
+}
+
+static void addEntityToCavity(FaceCavity* fc, apf::MeshEntity* e)
+{
+  PCU_ALWAYS_ASSERT(fc->mesh->getType(e) == apf::Mesh::TET);
+  fc->tets.push_back(e);
+}
+
+static void addEntitiesToCavity(
+    FaceCavity* fc, apf::DynamicArray<apf::MeshEntity*>& es)
+{
+  for (std::size_t i=0; i < es.getSize(); ++i)
+    addEntityToCavity(fc, es[i]);
+}
+
+static bool getInitialFaceCavity(FaceCavity* fc, apf::CavityOp* o)
+{
+  if (! o->requestLocality(&fc->entity, 1))
+    return false;
+
+  apf::DynamicArray<apf::MeshEntity*> adjacent;
+  fc->mesh->getAdjacent(fc->entity, 3, adjacent);
+  addEntitiesToCavity(fc, adjacent);
+  return true;
+}
+
+static bool buildFaceCavity(FaceCavity* fc, apf::CavityOp* o)
+{
+  if (!getInitialFaceCavity(fc, o)) return false;
+  return true;
+}
+
+
+static void computeCorrectedFlux(FaceCavity* fc)
+{
+  apf::MeshEntity* face = fc->entity;
+
+  // 1. get upward tets of the face
+  apf::Up up;
+  fc->mesh->getUp(face, up);
+  if (crv::isBoundaryEntity(fc->mesh, face))
+    PCU_ALWAYS_ASSERT(up.n == 1);
+  else
+    PCU_ALWAYS_ASSERT(up.n == 2);
+
+  apf::MeshEntity* firstTet = up.e[0];
+  apf::MeshEntity* secondTet = nullptr;
+  if (up.n == 2)
+    secondTet  = up.e[1];
+
+  // 2. get positions of the face in downward faces of upward tets
+  apf::Downward tet1_faces, tet2_faces;
+  int nf = fc->mesh->getDownward(firstTet, 2, tet1_faces);
+  if (up.n == 2)
+    fc->mesh->getDownward(secondTet, 2, tet2_faces);
+  int tet1_pos = -1;
+  int tet2_pos = -1;
+  tet1_pos = apf::findIn(tet1_faces, nf, face);
+  if (up.n == 2)
+    tet2_pos = apf::findIn(tet2_faces, nf, face);
+
+  // 3. get downward edges of the face
+  apf::Downward edges;
+  int nedges = fc->mesh->getDownward(face, 1, edges);
+
+  // 4. get theta coeffs on the face
+  double components[3];
+  apf::getComponents(fc->correctflux->theta, face, 0, components);
+  mth::Vector<double> theta_coeffs(3);
+  theta_coeffs(0) = components[0];
+  theta_coeffs(1) = components[1];
+  theta_coeffs(2) = components[2];
+
+
+  // 5. Evaluate and save corrected flux vector in an auxiliary field
+  int ftype = fc->mesh->getType(face);
+  PCU_ALWAYS_ASSERT(ftype == apf::Mesh::TRIANGLE);
+  int nfdofs = apf::countElementNodes(fc->correctflux->ef->getShape(), ftype);
+  apf::NewArray<apf::Vector3> vectorshapes(nfdofs);
+
+  apf::MeshElement* fme = apf::createMeshElement(fc->mesh, face);
```

Comment:
@Samiullah-Malik 
there is no corresponding `apf::destroyMeshElement(fme)` and 'apf::destroyElement(fel)' calls for these. They have to be added after you are done using these objects and before exiting this code block.

Reply:
Resolved.